flytekit/core/workflow.py

```diff
@@ -135,98 +163,114 @@ def name(self) -> str:
-            if len(output_names) != len(workflow_outputs):
-                raise Exception(f"Length mismatch {len(output_names)} vs {len(workflow_outputs)}")
-            for i, out in enumerate(output_names):
-                if isinstance(workflow_outputs[i], ConditionalSection):
-                    raise AssertionError("A Conditional block (if-else) should always end with an `else_()` clause")
-                t = self._native_interface.outputs[out]
-                b = binding_from_python_std(ctx, out, self.interface.outputs[out].type, workflow_outputs[i], t,)
-                bindings.append(b)
+        # Last is starting a local workflow execution
+        else:
+            # Run some sanity checks
+            # Even though the _local_execute call generally expects inputs to be Promises, we don't have to do the
+            # conversion here in this loop. The reason is because we don't prevent users from specifying inputs
+            # as direct scalars, which means there's another Promise-generating loop inside _local_execute too
+            for k, v in kwargs.items():
+                if k not in self.interface.inputs:
+                    raise ValueError(f"Received unexpected keyword argument {k}")
+                if isinstance(v, Promise):
+                    raise ValueError(f"Received a promise for a workflow call, when expecting a native value for {k}")
 
-        # Save all the things necessary to create an SdkWorkflow, except for the missing project and domain
-        self._nodes = all_nodes
-        self._output_bindings = bindings
-        if not output_names:
-            return None
-        if len(output_names) == 1:
-            return bindings[0]
-        return tuple(bindings)
+            with ctx.new_execution_context(mode=ExecutionState.Mode.LOCAL_WORKFLOW_EXECUTION) as ctx:
+                result = self._local_execute(ctx, **kwargs)
 
-    def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], Promise, VoidPromise]:
-        """
-        Performs local execution of a workflow. kwargs are expected to be Promises for the most part (unless,
-        someone has hardcoded in my_wf(input_1=5) or something).
-        :param ctx: The FlyteContext
-        :param kwargs: parameters for the workflow itself
-        """
-        logger.info(f"Executing Workflow {self._name}, ctx{ctx.execution_state.Mode}")
+            expected_outputs = len(self.python_interface.outputs)
+            if expected_outputs == 0:
+                if result is None or isinstance(result, VoidPromise):
+                    return None
+                else:
+                    raise Exception(f"Workflow local execution expected 0 outputs but something received {result}")
 
+            if (expected_outputs > 1 and len(result) == expected_outputs) or (
+                expected_outputs == 1 and result is not None
+            ):
+                if isinstance(result, Promise):
```

Comment:
perhaps a stupid q, but can a result be anything but a promise when expected_outputs == 1?

Reply:
Yeah, a workflow can be defined as:

```
def wf() -> int:
  t1()
  return 5
```